pkg/util/filesystem/concurrent_write.go

```diff
@@ -0,0 +1,169 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package filesystem
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cenkalti/backoff"
+	"github.com/gofrs/flock"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ArtifactBuilder is a generic interface for building, serializing, and deserializing artifacts.
+// The type parameter T represents the in-memory type of the artifact.
+type ArtifactBuilder[T any] interface {
+	Location() string
+	Generate() (T, error)
+	Serialize(T) ([]byte, error)
+	Deserialize([]byte) (T, error)
+	BackoffTimer() backoff.BackOff
+}
+
+// FetchArtifact attempts to load an artifact using the provided factory.
+// If the artifact does not exist, it return an error.
+func FetchArtifact[T any](factory ArtifactBuilder[T]) (T, error) {
+	location := factory.Location()
+	var zero T
+
+	// Read the artifact
+	content, err := os.ReadFile(location)
+	if err != nil {
+		return zero, fmt.Errorf("unable to read artifact: %s", err.Error())
+	}
+
+	// Try to load artifact
+	res, err := factory.Deserialize(content)
+	return res, err
+}
+
+func generateTmpArtifact[T any](factory ArtifactBuilder[T]) (T, string, error) {
```

Comment:
nitpick: I would move this method after `FetchOrCreateArtifact`. That way the order of the methods are public and then private 😄 I forces the reader to read the file from top to bottom